eventsources/sources/bitbucket/start.go

```diff
@@ -0,0 +1,344 @@
+
+	// When running multiple replicas of the eventsource, they will all try to create the webhook.
+	// Randomly sleep some time to mitigate the issue.
+	s1 := rand.NewSource(time.Now().UnixNano())
+	r1 := rand.New(s1)
+	time.Sleep(time.Duration(r1.Intn(2000)) * time.Millisecond)
+
+	err = router.saveBitbucketWebhook()
+	if err != nil {
+		logger.Errorw("failed to save Bitbucket webhook", zap.Error(err))
+	}
+
+	// Bitbucket hooks manager daemon
+	go func() {
+		// Another kind of race conditions might happen when pods do rolling upgrade - new pod starts
+		// and old pod terminates, if DeleteHookOnFinish is true, the hook will be deleted from Bitbucket.
+		// This is a workaround to mitigate the race conditions.
+		logger.Info("starting bitbucket hooks manager daemon")
+		ticker := time.NewTicker(60 * time.Second)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				logger.Info("exiting bitbucket hooks manager daemon")
+				return
+			case <-ticker.C:
+				err := router.saveBitbucketWebhook()
+				if err != nil {
+					logger.Errorw("failed to save Bitbucket webhook", zap.Error(err))
+				}
+			}
+		}
+	}()
+
+	return webhook.ManageRoute(ctx, router, controller, dispatch)
+}
+
+// chooseAuthStrategy returns an AuthStrategy based on the given credentials
+func (router *Router) chooseAuthStrategy() (AuthStrategy, error) {
+	es := router.bitbucketEventSource
+	switch {
+	case es.HasBitbucketBasicAuth():
+		return NewBasicAuthStrategy(es.Auth.Basic.Username, es.Auth.Basic.Password)
+	case es.HasBitbucketOAuthToken():
+		return NewOAuthTokenAuthStrategy(es.Auth.OAuthToken)
+	default:
+		return nil, errors.New("none of the supported auth options were provided")
+	}
+}
+
+// saveBitbucketWebhook creates or updates the configured webhook in Bitbucket
+func (router *Router) saveBitbucketWebhook() error {
+	logger := router.GetRoute().Logger
+	bitbucketEventSource := router.bitbucketEventSource
+	formattedWebhookURL := common.FormattedURL(bitbucketEventSource.Webhook.URL, bitbucketEventSource.Webhook.Endpoint)
+
+	logger.Info("listing existing webhooks...")
+	hooks, err := router.listWebhooks()
+	if err != nil {
+		logger.Errorw("failed to list webhooks", zap.Error(err))
+		return errors.Wrap(err, "failed to list webhooks")
+	}
+
+	logger.Info("checking if webhook already exists...")
+	existingHookSubscription, isFound := router.findWebhook(hooks, formattedWebhookURL)
+	if isFound && router.shouldUpdateWebhook(existingHookSubscription) {
+		logger.Info("webhook already exists but requires an update...")
+		if _, err = router.updateWebhook(existingHookSubscription); err != nil {
+			logger.Errorw("failed to update existing webhook", zap.Error(err))
+			return errors.Wrap(err, "failed to update existing webhook")
+		}
+
+		logger.Info("successfully updated the existing webhook")
+		return nil
+	}
+
+	logger.Info("webhook doesn't exist yet, creating a new webhook...")
+	newWebhook, err := router.createWebhook(formattedWebhookURL)
```

Comment:
shouldn't we check for just `if isFound` before creating a new one?

Reply:
Yes 100% good catch, thanks! I actually added the `shouldUpdateWebhook` check only later and forgot to handle this case.
Pushing a fix right now.